core/templates/domain/topic/topic-update.service.ts

```diff
@@ -22,719 +22,750 @@
-  'SUBTOPIC_PROPERTY_URL_FRAGMENT',
-  'TOPIC_PROPERTY_ABBREVIATED_NAME', 'TOPIC_PROPERTY_DESCRIPTION',
-  'TOPIC_PROPERTY_LANGUAGE_CODE', 'TOPIC_PROPERTY_META_TAG_CONTENT',
-  'TOPIC_PROPERTY_NAME', 'TOPIC_PROPERTY_PRACTICE_TAB_IS_DISPLAYED',
-  'TOPIC_PROPERTY_THUMBNAIL_BG_COLOR',
-  'TOPIC_PROPERTY_THUMBNAIL_FILENAME',
-  'TOPIC_PROPERTY_URL_FRAGMENT', function(
-      UndoRedoService,
-      CMD_ADD_SUBTOPIC, CMD_DELETE_ADDITIONAL_STORY,
-      CMD_DELETE_CANONICAL_STORY, CMD_DELETE_SUBTOPIC,
-      CMD_MOVE_SKILL_ID_TO_SUBTOPIC, CMD_REARRANGE_CANONICAL_STORY,
-      CMD_REARRANGE_SKILL_IN_SUBTOPIC,
-      CMD_REARRANGE_SUBTOPIC,
-      CMD_REMOVE_SKILL_ID_FROM_SUBTOPIC,
-      CMD_REMOVE_UNCATEGORIZED_SKILL_ID, CMD_UPDATE_SUBTOPIC_PAGE_PROPERTY,
-      CMD_UPDATE_SUBTOPIC_PROPERTY, CMD_UPDATE_TOPIC_PROPERTY,
-      SUBTOPIC_PAGE_PROPERTY_PAGE_CONTENTS_AUDIO,
-      SUBTOPIC_PAGE_PROPERTY_PAGE_CONTENTS_HTML,
-      SUBTOPIC_PROPERTY_THUMBNAIL_BG_COLOR,
-      SUBTOPIC_PROPERTY_THUMBNAIL_FILENAME, SUBTOPIC_PROPERTY_TITLE,
-      SUBTOPIC_PROPERTY_URL_FRAGMENT,
-      TOPIC_PROPERTY_ABBREVIATED_NAME, TOPIC_PROPERTY_DESCRIPTION,
-      TOPIC_PROPERTY_LANGUAGE_CODE, TOPIC_PROPERTY_META_TAG_CONTENT,
-      TOPIC_PROPERTY_NAME, TOPIC_PROPERTY_PRACTICE_TAB_IS_DISPLAYED,
-      TOPIC_PROPERTY_THUMBNAIL_BG_COLOR,
-      TOPIC_PROPERTY_THUMBNAIL_FILENAME,
-      TOPIC_PROPERTY_URL_FRAGMENT) {
-    // Creates a change using an apply function, reverse function, a change
-    // command and related parameters. The change is applied to a given
-    // topic.
-    // entity can be a topic object or a subtopic page object.
-    var _applyChange = function(entity, command, params, apply, reverse) {
-      var changeDict = angular.copy(params);
-      changeDict.cmd = command;
-      var changeObj = new Change(changeDict, apply, reverse);
-      UndoRedoService.applyChange(changeObj, entity);
-    };
-
-    var _getParameterFromChangeDict = function(changeDict, paramName) {
-      return changeDict[paramName];
-    };
-
-    // Applies a topic property change, specifically. See _applyChange()
-    // for details on the other behavior of this function.
-    var _applyTopicPropertyChange = function(
-        topic, propertyName, newValue, oldValue, apply, reverse) {
-      _applyChange(topic, CMD_UPDATE_TOPIC_PROPERTY, {
-        property_name: propertyName,
-        new_value: angular.copy(newValue),
-        old_value: angular.copy(oldValue) || null
-      }, apply, reverse);
-    };
-
-    var _applySubtopicPropertyChange = function(
-        topic, propertyName, subtopicId, newValue, oldValue, apply, reverse) {
-      _applyChange(topic, CMD_UPDATE_SUBTOPIC_PROPERTY, {
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+
+import cloneDeep from 'lodash/cloneDeep';
+
+import {
+  TopicRemoveSkillFromSubtopicChange,
+  TopicChange,
+  Change }
+  from 'domain/editor/undo_redo/change.model';
```

Comment:
Imports don't count towards max-len violations, you can move this into one line.